From 0a3bf1d7c830ef1c0e39c3a50648b3c0ea772ccc Mon Sep 17 00:00:00 2001
From: David Michael <david.michael@coreos.com>
Date: Thu, 20 Jul 2017 14:50:11 -0700
Subject: [PATCH] stage1-usr: Support local systemd source for offline builds

Local systemd source trees are useful here for packagers to build
usr_from_src without requiring network access.  This differentiates
between local and remote systemd sources by checking if the value
from ./configure contains "://".  If the string is found, the build
will preserve existing behavior.  If not, the value is interpreted
as a direct specification of the source directory path.
---
 configure.ac                        |  6 +++---
 stage1/usr_from_src/usr_from_src.mk | 20 ++++++++++++++++----
 2 files changed, 19 insertions(+), 7 deletions(-)

diff --git a/configure.ac b/configure.ac
index 60c4f84..8ee7fd3 100644
--- a/configure.ac
+++ b/configure.ac
@@ -105,7 +105,7 @@ AC_ARG_WITH(stage1-default-images-directory,
 
 AC_ARG_WITH([stage1-systemd-src],
             [AS_HELP_STRING([--with-stage1-systemd-src],
-                            [address to git repository of systemd, used in 'src' stage1 flavor (default: 'https://github.com/systemd/systemd.git')])],
+                            [URL to a systemd Git repository or path to a local systemd source directory, used in 'src' stage1 flavor (default: 'https://github.com/systemd/systemd.git')])],
             [RKT_STAGE1_SYSTEMD_SRC="${withval}"],
             [RKT_STAGE1_SYSTEMD_SRC='auto'])
 
@@ -869,8 +869,8 @@ RKT_IF_HAS_VALUE([${RKT_STAGE1_FLAVORS}], [src],
                   [AC_MSG_RESULT([
         src flavor specific build parameters
 
-        systemd git repo:                       '${RKT_STAGE1_SYSTEMD_SRC}'
-        systemd git revision:                   '${RKT_STAGE1_SYSTEMD_REV}'
+        systemd source:                         '${RKT_STAGE1_SYSTEMD_SRC}'
+        systemd Git revision:                   '${RKT_STAGE1_SYSTEMD_REV}'
         systemd version:                        '${RKT_STAGE1_SYSTEMD_VER}'])])
 
 RKT_IF_HAS_VALUE([${RKT_STAGE1_FLAVORS}], [kvm],
diff --git a/stage1/usr_from_src/usr_from_src.mk b/stage1/usr_from_src/usr_from_src.mk
index c73ca57..b7083e0 100644
--- a/stage1/usr_from_src/usr_from_src.mk
+++ b/stage1/usr_from_src/usr_from_src.mk
@@ -6,13 +6,20 @@
 $(call setup-tmp-dir,UFS_TMPDIR)
 # directory for systemd stuff (srcdir, builddir, installdir)
 UFS_SYSTEMDDIR := $(UFS_TMPDIR)/systemd
-# systemd src dir
-UFS_SYSTEMD_SRCDIR := $(UFS_SYSTEMDDIR)/src
 # systemd build dir
 UFS_SYSTEMD_BUILDDIR := $(UFS_SYSTEMDDIR)/build
 # systemd install dir, also serves as a temporary rootfs
 UFS_ROOTFSDIR := $(UFS_SYSTEMDDIR)/rootfs
 
+# systemd src dir
+ifeq ($(RKT_STAGE1_SYSTEMD_SRC),$(subst ://,,$(RKT_STAGE1_SYSTEMD_SRC)))
+# If the configured source does not appear to be a URL, use it as a path ...
+UFS_SYSTEMD_SRCDIR := $(RKT_STAGE1_SYSTEMD_SRC)
+else
+# ... otherwise, clone the URL with Git into this directory.
+UFS_SYSTEMD_SRCDIR := $(UFS_SYSTEMDDIR)/src
+endif
+
 # main stamp, makes sure that initial rootfs contents are prepared and
 # clean/deps/filelist are generated
 $(call setup-stamp-file,UFS_STAMP,systemd)
@@ -196,14 +203,14 @@ $(call generate-stamp-rule,$(UFS_SYSTEMD_CLONE_AND_PATCH_STAMP),$(UFS_SYSTEMD_CO
 # this patches the git repo and generates the configure script
 $(call forward-vars,$(UFS_SYSTEMD_CONFIGURE), \
 	UFS_PATCHES_DIR GIT UFS_SYSTEMD_SRCDIR UFS_AG_OUT)
-$(UFS_SYSTEMD_CONFIGURE):
+$(UFS_SYSTEMD_CONFIGURE): | $(UFS_TMPDIR)
 	$(VQ) \
 	set -e; \
 	shopt -s nullglob ; \
 	if [ -d "$(UFS_PATCHES_DIR)" ]; then \
 		for p in "$(abspath $(UFS_PATCHES_DIR))"/*.patch; do \
 			$(call vb,v2,PATCH,$${p#$(MK_TOPLEVEL_ABS_SRCDIR)/}) \
-			"$(GIT)" -C "$(UFS_SYSTEMD_SRCDIR)" am $(call vl3,--quiet) "$${p}"; \
+			patch $(call vl3,--silent )--directory="$(UFS_SYSTEMD_SRCDIR)" --strip=1 --forward <"$${p}"; \
 		done; \
 	fi; \
 	pushd "$(UFS_SYSTEMD_SRCDIR)" $(call vl3,>/dev/null); \
@@ -233,6 +240,9 @@ $(call generate-patches-filelist,$(UFS_PATCHES_FILELIST),$(UFS_PATCHES_DIR))
 # regenerated.
 $(call generate-glob-deps,$(UFS_PATCHES_DEPS_STAMP),$(UFS_SYSTEMD_CONFIGURE_AC),$(UFS_PATCHES_DEPMK),.patch,$(UFS_PATCHES_FILELIST),$(UFS_PATCHES_DIR),normal)
 
+# Clone the systemd source repository if not using a local source tree.
+ifneq ($(UFS_SYSTEMD_SRCDIR),$(RKT_STAGE1_SYSTEMD_SRC))
+
 # parameters for makelib/git.mk
 GCL_REPOSITORY := $(RKT_STAGE1_SYSTEMD_SRC)
 GCL_DIRECTORY := $(UFS_SYSTEMD_SRCDIR)
@@ -256,6 +266,8 @@ endif
 
 include makelib/git.mk
 
+endif
+
 # Remove the build directory if there were some changes in sources
 # (like different branch or repository, or a change in patches)
 $(UFS_SYSTEMD_RM_BUILDDIR_STAMP): $(UFS_SYSTEMD_CLONE_AND_PATCH_STAMP)
-- 
2.7.5

